### ISSUE TITLE
POST benchmark result: require at least one non-null iteration result?

### ISSUE THREAD

**Issue by jgehrcke** [conbench/conbench#813]
The current specification is unclear. Code is, too. Maybe this is covered by tests. 

This is a quick ticket based on https://github.com/conbench/conbench/pull/811/files#diff-9d75d1bf6ac23bfc056a410721cf4c9bccd8b302f8e3e83e4b4bf1698d0eaad7R99.

action item: properly specify this in text form (api spec), confirm that specified behavior is covered by tests, and also make sure that the business logic is readable, simple.

**Comment by jonkeane** [comment]
For posterity, the API docs that talk about this are at https://github.com/conbench/conbench/blob/aa95b6d52ff6cb6697b7ab8d04c43d6cf3612860/conbench/entities/benchmark_result.py#L239 (there's lots to improve about where+how this is surfaced, but I do not want to distract with that here!)

While I would love to live in a world where sending something like `[null, null, null]` as results was illegal (and effectively never happened), we've had experience with conbench users where they did, in fact, want to log if a benchmark entirely failed each time (along with a traceback in the error or optional benchmark info fields so that a developer can quickly see what has gone wrong and address it without needing to go into CI logs to find that traceback in the runner). 

Unless we decide to change this (though we will still need to support that all-iterations-errored case somehow), the API docs should be clearer that `[null, null, null]` is acceptable.

**Comment by jgehrcke** [comment]
Thank you for the feedback @jonkeane. This tech talk is fun.

>  we've had experience with conbench users where they did, in fact, want to log if a benchmark entirely failed each time

Sure. I think in that case one should send the wrapper `_BenchmarkFacadeSchemaCreate` without the `stats` property (which is optional):
```
stats = marshmallow.fields.Nested(BenchmarkResultSchema().create, required=False)
```


I'd love to have one recommended way for communicating the all-iterations-failed outcome. Between the two ways we offer today I think I like the do-not-set-stats-property more than the set-all-items-to-null.

In the outer structure one can use the `error` field to represent the wrongness in a custom way.

Let's also require `error` to be populated when no `stats` property was provided.

When there is at least one non-null iteration result but also at least one `null` result: we should not require `error` to be set, but document that we recommend to populate the `error` property with useful information about how the individual iteration(s) failed.

For the case where there is more than one failed iteration (but at least one non-null result) we have the challenge of having more than one error, but only one `error` struct. In that case we should recommend users to create their own iteration-error-mapping.

**Comment by jonkeane** [comment]
> For the case where there is more than one failed iteration (but at least one non-null result) we have the challenge of having more than one error, but only one error struct. In that case we should recommend users to create their own iteration-error-mapping.

Yeah, I think this is the best we can do (for now at least). It definitely is best practice to create an iteration-error-mapping, but enforcing that (without explicitly creating all that structure) is a lot to do. And these error info, tracebacks, etc. are really best-effort on our part: ideally they are used basically never, and as long as they've got decent(ish) info in them they unlock debugging pretty quickly. The ones I've seen in Arrow are almost always something like a segfault, which is enough to know that the problem is in the Arrow code, not the benchmark code (or at least that's very very unlikely!) versus something complaining about validation code or setting up the benchmark is much more likely the benchmark code. 

> I'd love to have one recommended way for communicating the all-iterations-failed outcome. Between the two ways we offer today I think I like the do-not-set-stats-property more than the set-all-items-to-null.

I'm totally fine to go this direction if it makes other things easier. I slightly prefer the send-all-nulls option since it speaks more to encourage folks to "create their own iteration-error-mapping" since they will always have to do that, no matter if they fail all or fail just one (in other words: no need to special case failing all iterations to do something special there). But my preference is only very slight there — I'm happy with either way. And most glad that we'll have this a bit less ambiguous too!

**Comment by jgehrcke** [comment]
> Yeah, I think this is the best we can do (for now at least).

:+1: 
Thanks for acking.

>  It definitely is best practice to create an iteration-error-mapping, but enforcing that (without explicitly creating all that structure) is a lot to do

Ack. I like the free JSON form. I am always in favor of having clear responsibility boundaries; it's lovely when a tiny and precise sentence in docs can move a lot of work elsewhere, without surprising the other end :-).

>  I slightly prefer the send-all-nulls option since it speaks more to encourage folks to "create their own iteration-error-mapping" since they will always have to do that

We can also make the `stats` property required, and also _require_ the `error` property to be populated in case there is no non-null value in `data`/`times`.


My current mental model:

## variant A

- when no `stats` property is set, the `error` property must be set to a non-empty dictionary (otherwise the request is rejected).
- when `stats` are provided
    - `data` must be an array of non-zero length 
    - `data` must contain at least one non-null value
    - if `data` contains at least one null value then the `BenchmarkResult` is considered failed. If no `error` information was provided by the user, then a dummy error message "not all iterations completed" is automatically set by Conbench upon DB insert.
    
## variant B

- require the `stats` property.
- `data` must be an array of non-zero length
- `data` is allowed to have only null values. In this case, the `error` property must be set to a non-empty dictionary (otherwise the request is rejected).
- if `data` contains at least one null value then the `BenchmarkResult` is considered failed. If no `error` information was provided by the user, then a dummy error message "not all iterations completed" is automatically set by Conbench upon DB insert.

Variant A might have the advantage that in some real-world scenarios when no iteration was executed (for the error reason) it becomes ~easy to communicate the error information.

When we opt for variant B we run into what I think are weirdnesses in the `BenchmarkResultCreate` schema (which really should be called something with `...Stats...`. As of today it _requires_ the iterations count (what explodes if it does not match length of `data`?), it also requires the presence of the `times` property (but then doesn't apply any rules to that). So, if everything failed and no iteration was executed, transmitting the error information requires the client to create a data structure that doesn't look right (this communicates more than just error information, it also communicates "useless" info):
```python
stats = {
    iterations: 3,
    data: [null, null, null]
    times: []
}
```

If we do variant B, we should maybe allow for setting

```python
stats = {
    data: [null, null, null]
}
```


Moving this aspect of the discussion to https://github.com/conbench/conbench/issues/823.

**Comment by alistaire47** [comment]
Most of our producers of results aren't yet populating failed iterations as nulls because they were written before #416. Most of them still report the whole result as a failure if one iteration fails and don't send `stats.data`, as the schema use to strictly require `error` xor `stats`.

benchadapt should be able to accept and send partially-failed results, though most of the adapted things aren't set up to pass failures; they just explode. benchrun has some code around this (but nothing is using it). arrowbench and labs/benchmarks (and the conbench runner) are all-or-nothing.

To update all these to bring them in line with the ideas here, there is in some cases some non-trivial work to handle iteration failures and reporting partial failures. We're not going to fix this for all of these (certainly not labs/benchmarks), but we should at least make stories where we want to (regardless of when we get to it).

**Comment by jonkeane** [comment]
> Variant A might have the advantage that in some real-world scenarios when no iteration was executed (for the error reason) it becomes ~easy to communicate the error information.

Yeah, I agree with this: it does make that easier. Also in light of what @alistaire47 mentions too, Variant A should be more compatible with the runners we have on the ground right now (which is good, we shouldn't bite off a huge amount of change work here if we don't otherwise have to — I strongly agree that consistency is important, but if we need to do lots and lots of work to prepare for that, we should reassess if now is the time for that consistency enforcement). 

We might further refine Variant A:

* when no stats property is set, the error property must be set to a non-empty dictionary (otherwise the request is rejected).
when stats are provided
    * data must be an array of non-zero length
    * ~data must contain at least one non-null value~
    * if data contains at least one null value then the BenchmarkResult is considered failed. If no error information was provided by the user, then a dummy error message "not all iterations completed" is automatically set by Conbench upon DB insert.
    
Which would allow for cases where no iteration ran (`error = {"msg"="something really bad happened in setup"}`, `data = null`) as well as there were iterations, but each one failed (`error = [{"msg"="something bad happened in the query"}, {"msg"="something bad happened in the query"}, {"msg"="something bad happened in the query"}]`, `data = [null, null, null]`). Maybe that's too much nuance and no one would ever need it?

**Comment by jgehrcke** [comment]
From PFE-901 (internal), a related question:

> For every hash I see three bars, which I assume is for the three iterations, but I always see either all three or none. Does that mean that if one of the three iterations fails, then we dont collect any of the benchmark info and we just label it all as failure?

CC @wmalpica
